### PR TITLE
set_reserved_byte_for_start-of-stream-SEI

### PR DESCRIPTION
**Pull request by mithydolphin**
### Describe your changes

Please include a summary of the change, a relevant motivation and context.

### Issue ticket number and link

- Fixes #(issue)

### Checklist before requesting a review

- [ ] I have performed a self-review of my own code
- [ ] I have verified that the code builds perfectly fine on my local system
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have verified that my code follows the style already available in the repository
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: tests/check/check_h26xsigned_auth.c [hunk 1]
```diff
@@ -2061,6 +2062,82 @@ START_TEST(no_public_key_in_sei_and_bad_public_key_on_validation_side)
 }
 END_TEST
 
+/* Test description
+ * 1. Setup a signed_video_t session
+ * 2. Set |is_start_stream| flag as true
+ * 3. Add 1 P NALU and 2 I NALU for signing that will trigger 2 SEI NALU
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_auth.c:2068]: This test is not really setup as a correct use case. BUT this gives me a good idea for how this can (should) work.
I think it should be possible for a user to (in pseudo code) do
```
sv = signed_video_create(codec);
signed_video_enable_use_of_start_of_stream_sei(sv);
signed_video_generate_start_of_stream_sei(sv);
// Add NAL Units for signing and fetch SEIs as normal. The first one will be a "golden SEI"
```

But this should be tested in a different PR, when more parts are in place. I suggest that you instead of this test should add a line in one of the helper functions that does
```
sv_rc = signed_video_is_start_of_stream(sv, nalu_data, nalu_data_size);
ck_assert_int_eq(sv_rc, SV_OK);
```

File: tests/check/check_h26xsigned_auth.c [hunk 2]
```diff
@@ -30,6 +30,7 @@
 #ifdef SV_VENDOR_AXIS_COMMUNICATIONS
 #include "lib/src/includes/sv_vendor_axis_communications.h"
 #endif
+#include "lib/src/signed_video_h26x_internal.h"  // h26x_nalu_t
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_auth.c:33]: I don't think this is used anywhere

File: lib/src/signed_video_h26x_common.c [hunk 3]
```diff
@@ -1312,3 +1313,15 @@ signed_video_compare_versions(const char *version1, const char *version2)
 error:
   return status;
 }
+
+bool
+signed_video_is_start_of_stream(uint8_t *sei, size_t sei_size, SignedVideoCodec codec)
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_common.c:1318]: I think it is better to pass in a `signed_video_t` than the `codec` and make this function public in signed_video_auth.h so the validation side can use it.
Also, rename it to `signed_video_is_start_of_stream_sei` and replace sei -> nalu, since it should be possible to pass in any type of NAL Unit.

File: lib/src/signed_video_h26x_common.c [hunk 4]
```diff
@@ -1312,3 +1313,15 @@ signed_video_compare_versions(const char *version1, const char *version2)
...
+{
+  h26x_nalu_t nalu = parse_nalu_info(sei, sei_size, codec, false, true);
+  if (nalu.nalu_type != NALU_TYPE_SEI) return false;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_common.c:1321]: Isn't it enough to simplify this function to
```
h26x_nalu_t nalu = parse_nalu_info(sei, sei_size, codec, false, true);
return nalu.start_of_sei;
```
Plus adding sanity checks on inputs.

File: lib/src/includes/signed_video_sign.h [hunk 5]
```diff
@@ -519,6 +519,8 @@ signed_video_set_max_sei_payload_size(signed_video_t *self, size_t max_sei_paylo
 SignedVideoReturnCode
 signed_video_set_hash_algo(signed_video_t *self, const char *name_or_oid);
 
+SignedVideoReturnCode
+signed_video_set_is_start_stream(signed_video_t *self, bool is_start_stream);
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_sign.h:523]: I think this belongs to a different PR, where usage of golden SEIs are applied.

**Review by bjornvolcker — CHANGES_REQUESTED** [review]
Good start. We can discuss what to test in this PR and the upcoming ones.

File: lib/src/signed_video_internal.h [hunk 6]
```diff
@@ -123,6 +123,8 @@ struct _signed_video_t {
   SignedVideoAuthenticityLevel authenticity_level;
   bool add_public_key_to_sei;
   bool sei_epb;  // Flag that tells whether to generate SEI frames w/wo emulation prevention bytes
+  bool is_start_stream;  // Flag that tells whether to generate SEI frames w/wo start_stream_sei
+  uint8_t *reserved_byte;
```
[the remark this answers is not in the document]
Comment (mithydolphin) [reply lib/src/signed_video_internal.h:127]: It is used in generate_sei_nalu while setting to reserved bytes

[the remark this answers is not in the document]
Comment (bjornvolcker) [reply lib/src/signed_video_internal.h:127]: I can only see that it is being set there, but it is then never used.

File: lib/src/signed_video_h26x_common.c [hunk 7]
```diff
@@ -583,10 +583,10 @@ remove_epb_from_sei_payload(h26x_nalu_t *nalu)
   nalu->tlv_start_in_nalu_data = nalu->payload + UUID_LEN;
   nalu->tlv_size = nalu->payload_size - UUID_LEN;
   nalu->reserved_byte = *nalu->tlv_start_in_nalu_data;
-  nalu->start_of_sei = ((nalu->reserved_byte & (1 << 6)) != 0);
   nalu->tlv_start_in_nalu_data++;  // Move past the |reserved_byte|.
   nalu->tlv_size -= 1;  // Exclude the |reserved_byte| from TLV size.
   nalu->with_epb = (nalu->reserved_byte & 0x80);  // Hash with emulation prevention bytes
+  nalu->is_start_of_stream_sei = (nalu->reserved_byte & 0x40);  // Parse flags from reserved byte
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_common.c:589]: The comment came out wrong. This comment should be placed before all parsed flags. There will be more flags coming up in the future.

File: tests/check/check_h26xsigned_auth.c [hunk 8]
```diff
@@ -2094,47 +2089,28 @@ START_TEST(_is_start_of_stream_byte)
...
 
   // Add I-frame for signing and get SEI frame.
   sv_rc = signed_video_add_nalu_for_signing_with_timestamp(
       sv, i_1_nalu->data, i_1_nalu->data_size, &g_testTimestamp);
   ck_assert_int_eq(sv_rc, SV_OK);
-  // Check the first I frame
+  // Check the I frame
   ck_assert_int_eq(
-      signed_video_is_start_of_stream(i_1_nalu->data, i_1_nalu->data_size, codec), false);
+      signed_video_is_start_of_stream_sei(sv, i_1_nalu->data, i_1_nalu->data_size), false);
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_auth.c:2099]: You should remove this entire test. It adds no value. It is better to add this type of line in an appropriate function in signed_video_helpers.c

File: lib/src/includes/signed_video_sign.h [hunk 9]
```diff
@@ -519,8 +519,8 @@ signed_video_set_max_sei_payload_size(signed_video_t *self, size_t max_sei_paylo
 SignedVideoReturnCode
 signed_video_set_hash_algo(signed_video_t *self, const char *name_or_oid);
 
-SignedVideoReturnCode
-signed_video_set_is_start_stream(signed_video_t *self, bool is_start_stream);
+bool
+signed_video_is_start_of_stream_sei(signed_video_t *self, uint8_t *sei, size_t sei_size);
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_sign.h:523]: This declaration should go into signed_video_auth.h since it is aimed towards the validation side. If you think it adds value to the signing side as well you should put it in signed_video_common.h

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

File: lib/src/includes/signed_video_common.h [hunk 10]
```diff
@@ -123,6 +123,18 @@ signed_video_get_version();
 int
 signed_video_compare_versions(const char* version1, const char* version2);
 
+/**
+ * @brief Checks the NALU if its a golden SEI
+ *
+ * @param sei SEI payload
+ * @param sei_size Size of the SEI NALU
+ *
+ * @returns True if NALU is the golden SEI
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_common.h:132]: grammar: the -> a

File: lib/src/includes/signed_video_common.h [hunk 11]
```diff
@@ -123,6 +123,18 @@ signed_video_get_version();
...
+ *          False if NALU is not the golden SEI
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_common.h:133]: Redundant information since you return a boolean

File: lib/src/signed_video_authenticity.c [hunk 12]
```diff
@@ -398,3 +398,10 @@ signed_video_authenticity_report_free(signed_video_authenticity_t *authenticity_
 
   free(authenticity_report);
 }
+
+_Bool
+signed_video_is_golden_sei(signed_video_t *self, uint8_t *sei, unsigned long sei_size)
```
Comment (bjornvolcker) [inline lib/src/signed_video_authenticity.c:403]: This declaration should be placed in signed_video_h26x_common.c

File: lib/src/signed_video_h26x_common.c [hunk 13]
```diff
@@ -586,6 +586,7 @@ remove_epb_from_sei_payload(h26x_nalu_t *nalu)
   nalu->tlv_start_in_nalu_data++;  // Move past the |reserved_byte|.
   nalu->tlv_size -= 1;  // Exclude the |reserved_byte| from TLV size.
   nalu->with_epb = (nalu->reserved_byte & 0x80);  // Hash with emulation prevention bytes
+  nalu->is_golden_sei = (nalu->reserved_byte & 0x40);  // The NALU is golden SEI.
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_common.c:589]: grammar: "... is a golden..."

File: lib/src/signed_video_h26x_internal.h [hunk 14]
```diff
@@ -197,4 +198,6 @@ copy_nalu_except_pointers(h26x_nalu_t *dst_nalu, const h26x_nalu_t *src_nalu);
 void
 update_hashable_data(h26x_nalu_t *nalu);
 
+bool
+signed_video_is_start_of_stream(uint8_t *sei, size_t sei_size, SignedVideoCodec codec);
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_internal.h:202]: I think you forgot to remove this declaration.

File: lib/src/signed_video_h26x_sign.c [hunk 15]
```diff
@@ -290,7 +290,10 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
     h26x_set_nal_uuid_type(self, &payload_ptr, UUID_TYPE_SIGNED_VIDEO);
 
     // Add reserved byte(s).
+    // The bit stream is illustrated below.
+    // reserved_byte = |sei_epb|is_start_stream|0|0|0|0|0|0|
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:294]: `|epb|golden sei|0|0|0|0|0|0|`

File: lib/src/signed_video_internal.h [hunk 16]
```diff
@@ -123,6 +123,7 @@ struct _signed_video_t {
   SignedVideoAuthenticityLevel authenticity_level;
   bool add_public_key_to_sei;
   bool sei_epb;  // Flag that tells whether to generate SEI frames w/wo emulation prevention bytes
+  bool is_start_stream;  // Flag that tells whether to generate SEI frames w/wo start_stream_sei
```
Comment (bjornvolcker) [inline lib/src/signed_video_internal.h:126]: Rename to `is_golden_sei`

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

File: lib/src/includes/signed_video_common.h [hunk 17]
```diff
@@ -21,7 +21,9 @@
 
 #ifndef __SIGNED_VIDEO_COMMON_H__
 #define __SIGNED_VIDEO_COMMON_H__
+
 #include <stdbool.h>
+#include <stddef.h>
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_common.h:26]: I think you also need stdint.h for uint8_t

File: lib/src/includes/signed_video_common.h [hunk 18]
```diff
@@ -127,16 +129,16 @@ signed_video_compare_versions(const char* version1, const char* version2);
 /**
  * @brief Checks the NALU if its a golden SEI
  *
- * Golden SEI is the NALU that is generated at the start of the stream. This SEI includes only
- * the TLV tags which are only needed once.
+ * Golden SEI is the NALU that is generated at the start of the stream. This SEI is self-signed,
+ * hence not bound to a stream and it includes only the TLV tags which are only needed once.
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_common.h:133]: grammar: ... not bound to a stream, and it only includes information needed once, such as the Public key.

File: tests/check/signed_video_helpers.c [hunk 19]
```diff
@@ -134,6 +134,7 @@ create_signed_nalus_with_sv(signed_video_t *sv, const char *str, bool split_nalu
     } else {
       rc = signed_video_add_nalu_part_for_signing_with_timestamp(
           sv, item->data, item->data_size, &g_testTimestamp, true);
+      ck_assert_int_eq(signed_video_is_golden_sei(sv, item->data, item->data_size), false);
```
Comment (bjornvolcker) [inline tests/check/signed_video_helpers.c:137]: Since signed_video_is_golden_sei() returns a boolean I think it is better to use ck_assert().

Also, add a comment why this is test is applied here and not if the NALU is split.

**Review by bjornvolcker — CHANGES_REQUESTED** [review]
There are still unresolved review comments

File: lib/src/signed_video_h26x_common.c [hunk 20]
```diff
@@ -1312,3 +1313,10 @@ signed_video_compare_versions(const char *version1, const char *version2)
 error:
   return status;
 }
+
+bool
+signed_video_is_golden_sei(signed_video_t *self, const uint8_t *sei, size_t sei_size)
+{
+  h26x_nalu_t nalu = parse_nalu_info(sei, sei_size, self->codec, false, true);
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_common.c:1320]: You will get a segmentation fault here if `self` is a null pointer. Since this is a public API, there should be sanity checks on the input parameters.

**Review by bjornvolcker — CHANGES_REQUESTED** [review]
There are still old review comments to address

File: lib/src/signed_video_h26x_common.c [hunk 21]
```diff
@@ -1317,6 +1317,8 @@ signed_video_compare_versions(const char *version1, const char *version2)
 bool
 signed_video_is_golden_sei(signed_video_t *self, const uint8_t *sei, size_t sei_size)
 {
+  if (!self) return false;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_common.c:1320]: What about the other input parameters?

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

File: tests/check/signed_video_helpers.c [hunk 22]
```diff
@@ -134,6 +134,8 @@ create_signed_nalus_with_sv(signed_video_t *sv, const char *str, bool split_nalu
     } else {
       rc = signed_video_add_nalu_part_for_signing_with_timestamp(
           sv, item->data, item->data_size, &g_testTimestamp, true);
+      // The check is only performed for the complete NALU.
+      ck_assert(!signed_video_is_golden_sei(sv, item->data, item->data_size));
```
Comment (bjornvolcker) [inline tests/check/signed_video_helpers.c:138]: I just realized that the nalu split is done explicitly, so this call should be done before the if-statement.

**Comment by bjornvolcker** [comment]
There is a memory leak somewhere. You should run valgrind locally to get a callstack
`$ CK_FORK=no valgrind --leak-check=full build/tests/check/unittest_sign`

**Review by bjornvolcker — APPROVED** [review]
